src/ring_theory/principal_ideal_domain.lean

```diff
@@ -6,55 +6,64 @@ Author: Chris Hughes, Morenikeji Neri
 import ring_theory.noetherian
 import ring_theory.unique_factorization_domain
 
-variables {α : Type*}
+universes u v
+variables {R : Type u} {M : Type v}
 
-open set function ideal
+open set function
+open submodule
 open_locale classical
 
-class ideal.is_principal [comm_ring α] (S : ideal α) : Prop :=
-(principal [] : ∃ a, S = span {a})
+/-- An `R`-submodule of `M` is principal if it is generated by one element. -/
+class submodule.is_principal [ring R] [add_comm_group M] [module R M] (S : submodule R M) : Prop :=
```

Comment:
Should this actually be a class, in hindsight?

Reply:
I wouldn't make it a class if I wrote it today, but I don't have any further argumentation for that. I guess it's nicer to write `generator I` than `generator (ℤ.is_principal.principal I)`, but we could just use `[fact (is_principal I)]` instead. There don't seem to be any issues in this file with keeping it as a class. My suggestion: change it to a `def` if those issues start appearing, e.g. with the Dedekind domain development.